File: src/routes/customers.py
"""Customer routes: /api/customers"""
from flask import Blueprint, request, jsonify, g
from src.db.init_db import get_db
from src.middleware.auth import require_auth, require_permission

customers_bp = Blueprint("customers", __name__, url_prefix="/api/customers")


def _row_to_dict(row):
    return dict(row) if row else None


@customers_bp.get("/")
@require_auth
@require_permission("VIEW_CUSTOMERS")
def list_customers():
    db = get_db()
    page = max(1, request.args.get("page", 1, type=int))
    per_page = min(100, request.args.get("per_page", 20, type=int))
    search = request.args.get("search", "").strip()
    offset = (page - 1) * per_page

    base_q = """
        FROM customer c
        LEFT JOIN theme t ON c.accnt_theme_id = t.theme_id
    """
    where = ""
    params = []
    if search:
        where = "WHERE c.first_name LIKE ? OR c.last_name LIKE ? OR c.email LIKE ?"
        s = f"%{search}%"
        params = [s, s, s]

    total = db.execute(f"SELECT COUNT(*) {base_q} {where}", params).fetchone()[0]
    rows = db.execute(
        f"""SELECT c.*, t.theme_name {base_q} {where}
            ORDER BY c.customer_id LIMIT ? OFFSET ?""",
        params + [per_page, offset]
    ).fetchall()
    db.close()

    return jsonify({
        "data": [dict(r) for r in rows],
        "pagination": {"page": page, "per_page": per_page, "total": total,
                       "pages": (total + per_page - 1) // per_page}
    })


@customers_bp.get("/<int:customer_id>")
@require_auth
@require_permission("VIEW_CUSTOMERS")
def get_customer(customer_id):
    db = get_db()
    customer = db.execute(
        """SELECT c.*, t.theme_name FROM customer c
           LEFT JOIN theme t ON c.accnt_theme_id = t.theme_id
           WHERE c.customer_id = ?""",
        (customer_id,)
    ).fetchone()

    if not customer:
        db.close()
        return jsonify({"error": "Customer not found"}), 404

    addresses = db.execute(
        "SELECT * FROM address WHERE acc_id = ?", (customer_id,)
    ).fetchall()

    billing = db.execute(
        "SELECT * FROM billing_address WHERE acc_id = ?", (customer_id,)
    ).fetchall()

    themes = db.execute("""
        SELECT t.theme_id, t.theme_name FROM customer_theme ct
        JOIN theme t ON ct.theme_id = t.theme_id
        WHERE ct.customer_id = ?
    """, (customer_id,)).fetchall()

    ratings = db.execute("""
        SELECT cr.rating_id, cr.rating_name FROM customer_content_rating ccr
        JOIN content_rating cr ON ccr.rating_id = cr.rating_id
        WHERE ccr.customer_id = ?
    """, (customer_id,)).fetchall()

    subscriptions = db.execute("""
        SELECT s.*, sp.plan_name, sp.price FROM subscription s
        JOIN subscription_plan sp ON s.plan_id = sp.plan_id
        WHERE s.cus_id = ?
        ORDER BY s.start_date DESC
    """, (customer_id,)).fetchall()

    db.close()

    return jsonify({
        **dict(customer),
        "addresses": [dict(a) for a in addresses],
        "billing_addresses": [dict(b) for b in billing],
        "theme_preferences": [dict(t) for t in themes],
        "content_rating_preferences": [dict(r) for r in ratings],
        "subscriptions": [dict(s) for s in subscriptions],
    })


@customers_bp.post("/")
@require_auth
@require_permission("EDIT_CUSTOMERS")
def create_customer():
    data = request.get_json(silent=True) or {}
    required = ["first_name", "last_name", "email", "birth_date", "accnt_theme_id"]
    missing = [f for f in required if not data.get(f)]
    if missing:
        return jsonify({"error": f"Missing required fields: {', '.join(missing)}"}), 400

    db = get_db()
    try:
        cur = db.execute("""
            INSERT INTO customer (first_name, last_name, email, phone_number, birth_date,
                age_restricted, accnt_theme_id, clothing_size, age_rating_pref)
            VALUES (?,?,?,?,?,?,?,?,?)
        """, (
            data["first_name"], data["last_name"], data["email"],
            data.get("phone_number"), data["birth_date"],
            int(data.get("age_restricted", 0)), data["accnt_theme_id"],
            data.get("clothing_size"), data.get("age_rating_pref", "ALL")
        ))
        db.commit()
        customer_id = cur.lastrowid
    except Exception as e:
        db.close()
        if "UNIQUE" in str(e):
            return jsonify({"error": "Email already exists"}), 409
        return jsonify({"error": str(e)}), 400

    db.close()
    return jsonify({"message": "Customer created", "customer_id": customer_id}), 201


@customers_bp.put("/<int:customer_id>")
@require_auth
@require_permission("EDIT_CUSTOMERS")
def update_customer(customer_id):
    data = request.get_json(silent=True) or {}
    db = get_db()
    existing = db.execute("SELECT 1 FROM customer WHERE customer_id = ?", (customer_id,)).fetchone()
    if not existing:
        db.close()
        return jsonify({"error": "Customer not found"}), 404

    allowed = ["first_name", "last_name", "email", "phone_number",
               "clothing_size", "age_rating_pref", "accnt_theme_id", "age_restricted"]
    updates = {k: v for k, v in data.items() if k in allowed}

    if not updates:
        db.close()
        return jsonify({"error": "No valid fields to update"}), 400

    set_clause = ", ".join(f"{k} = ?" for k in updates)
    values = list(updates.values()) + [customer_id]

    try:
        db.execute(f"UPDATE customer SET {set_clause} WHERE customer_id = ?", values)
        db.execute(
            "INSERT INTO audit_log (user_id,entity_name,entity_id,action_type,old_value,new_value) VALUES (?,?,?,?,?,?)",
            (g.user["user_id"], "customer", customer_id, "UPDATE", "fields updated", str(updates))
        )
        db.commit()
    except Exception as e:
        db.close()
        return jsonify({"error": str(e)}), 400

    db.close()
    return jsonify({"message": "Customer updated"})


@customers_bp.delete("/<int:customer_id>")
@require_auth
@require_permission("DELETE_CUSTOMERS")
def delete_customer(customer_id):
    db = get_db()
    existing = db.execute("SELECT 1 FROM customer WHERE customer_id = ?", (customer_id,)).fetchone()
    if not existing:
        db.close()
        return jsonify({"error": "Customer not found"}), 404

    # Soft delete via deactivation isn't in the schema, so we check for dependencies
    has_subs = db.execute("SELECT 1 FROM subscription WHERE cus_id = ?", (customer_id,)).fetchone()
    if has_subs:
        db.close()
        return jsonify({"error": "Cannot delete customer with active subscriptions"}), 409

    db.execute("DELETE FROM customer_theme WHERE customer_id = ?", (customer_id,))
    db.execute("DELETE FROM customer_content_rating WHERE customer_id = ?", (customer_id,))
    db.execute("DELETE FROM address WHERE acc_id = ?", (customer_id,))
    db.execute("DELETE FROM billing_address WHERE acc_id = ?", (customer_id,))
    db.execute("DELETE FROM customer WHERE customer_id = ?", (customer_id,))
    db.commit()
    db.close()
    return jsonify({"message": "Customer deleted"})

File: functions/src/routes/customers.ts
import { Router, Response } from "express";
import * as admin from "firebase-admin";
import { FieldValue } from "firebase-admin/firestore";
import { db } from "../index";
import { requireAuth, requirePermission, AuthRequest, writeAudit } from "../middleware/auth";

export const customersRouter = Router();

/** Find the customer doc for the current auth user (by auth_uid, then email fallback) */
async function findMyCustomer(uid: string, email?: string) {
  // Try auth_uid first
  let snap = await db.collection("customers")
    .where("auth_uid", "==", uid).limit(1).get();
  if (!snap.empty) return snap.docs[0];

  // Fallback: match by email (for accounts created before auth_uid was added)
  if (email) {
    snap = await db.collection("customers")
      .where("email", "==", email).limit(1).get();
    if (!snap.empty) {
      // Backfill auth_uid so future lookups are fast
      await snap.docs[0].ref.update({ auth_uid: uid });
      return snap.docs[0];
    }
  }
  return null;
}

/** GET /api/customers — paginated list with optional search */
customersRouter.get(
  "/",
  requireAuth,
  requirePermission("VIEW_CUSTOMERS"),
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const pageSize = Math.min(100, parseInt(req.query.per_page as string) || 20);
      const search   = (req.query.search as string || "").toLowerCase();

      let query: admin.firestore.Query = db.collection("customers").orderBy("last_name");

      // Firestore doesn't support full-text search natively — use startAt/endAt for prefix
      if (search) {
        query = query
          .where("search_index", "array-contains", search);
      }

      // Cursor-based pagination
      const afterDoc = req.query.after as string | undefined;
      if (afterDoc) {
        const snap = await db.collection("customers").doc(afterDoc).get();
        if (snap.exists) query = query.startAfter(snap);
      }

      const snapshot = await query.limit(pageSize).get();
      const customers = snapshot.docs.map((d) => ({ id: d.id, ...d.data() }));

      res.json({
        data: customers,
        next_cursor: snapshot.docs.length === pageSize
          ? snapshot.docs[snapshot.docs.length - 1].id
          : null,
      });
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);

/** GET /api/customers/me — get own profile by auth UID */
customersRouter.get(
  "/me",
  requireAuth,
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const customerDoc = await findMyCustomer(req.user!.uid, req.user!.email);
      if (!customerDoc) {
        res.status(404).json({ error: "No customer profile found" }); return;
      }
      const addresses = await db.collection("customers")
        .doc(customerDoc.id).collection("addresses").get();

      res.json({
        id: customerDoc.id,
        ...customerDoc.data(),
        addresses: addresses.docs.map((d) => ({ id: d.id, ...d.data() })),
      });
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);

/** PUT /api/customers/me — update own profile */
customersRouter.put(
  "/me",
  requireAuth,
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const customerDoc = await findMyCustomer(req.user!.uid, req.user!.email);
      if (!customerDoc) {
        res.status(404).json({ error: "No customer profile found" }); return;
      }
      const allowed = ["first_name", "last_name", "phone_number",
                       "clothing_size", "age_rating_pref", "accnt_theme_id"];
      const updates: Record<string, unknown> = {};
      for (const key of allowed) {
        if (req.body[key] !== undefined) updates[key] = req.body[key];
      }
      if (Object.keys(updates).length === 0) {
        res.status(400).json({ error: "No valid fields to update" }); return;
      }
      updates.updated_at = FieldValue.serverTimestamp();
      await db.collection("customers").doc(customerDoc.id).update(updates);
      res.json({ message: "Profile updated" });
    } catch (err) {
      res.status(400).json({ error: String(err) });
    }
  }
);

/** DELETE /api/customers/me — delete own account */
customersRouter.delete(
  "/me",
  requireAuth,
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const customerDoc = await findMyCustomer(req.user!.uid, req.user!.email);
      if (!customerDoc) {
        res.status(404).json({ error: "No customer profile found" }); return;
      }
      const hasSubs = await db.collection("subscriptions")
        .where("customer_id", "==", req.user!.uid).limit(1).get();
      if (!hasSubs.empty) {
        res.status(409).json({ error: "Cannot delete account with active subscriptions. Cancel them first." }); return;
      }
      // Delete addresses subcollection
      const addresses = await db.collection("customers").doc(customerDoc.id).collection("addresses").get();
      const batch = db.batch();
      addresses.docs.forEach((d) => batch.delete(d.ref));
      batch.delete(customerDoc.ref);
      await batch.commit();
      res.json({ message: "Account deleted" });
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);

/** POST /api/customers/me/addresses — add address */
customersRouter.post(
  "/me/addresses",
  requireAuth,
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const customerDoc = await findMyCustomer(req.user!.uid, req.user!.email);
      if (!customerDoc) { res.status(404).json({ error: "No customer profile found" }); return; }
      const customerId = customerDoc.id;
      const { label, street, city, province, postal_code, country, is_default } = req.body;
      if (!street || !city || !province || !postal_code || !country) {
        res.status(400).json({ error: "Missing required address fields" }); return;
      }
      // If this is default, unset other defaults
      if (is_default) {
        const existing = await db.collection("customers").doc(customerId)
          .collection("addresses").where("is_default", "==", true).get();
        const batch = db.batch();
        existing.docs.forEach((d) => batch.update(d.ref, { is_default: false }));
        await batch.commit();
      }
      const ref = await db.collection("customers").doc(customerId).collection("addresses").add({
        label: label || "Home",
        street, city, province, postal_code, country,
        is_default: is_default ?? false,
        created_at: FieldValue.serverTimestamp(),
      });
      res.status(201).json({ message: "Address added", address_id: ref.id });
    } catch (err) {
      res.status(400).json({ error: String(err) });
    }
  }
);

/** PUT /api/customers/me/addresses/:addrId — update address */
customersRouter.put(
  "/me/addresses/:addrId",
  requireAuth,
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const customerDoc = await findMyCustomer(req.user!.uid, req.user!.email);
      if (!customerDoc) { res.status(404).json({ error: "No customer profile found" }); return; }
      const customerId = customerDoc.id;
      const addrRef = db.collection("customers").doc(customerId)
        .collection("addresses").doc(req.params.addrId);
      const addrDoc = await addrRef.get();
      if (!addrDoc.exists) { res.status(404).json({ error: "Address not found" }); return; }
      const allowed = ["label", "street", "city", "province", "postal_code", "country", "is_default"];
      const updates: Record<string, unknown> = {};
      for (const key of allowed) {
        if (req.body[key] !== undefined) updates[key] = req.body[key];
      }
      if (req.body.is_default) {
        const existing = await db.collection("customers").doc(customerId)
          .collection("addresses").where("is_default", "==", true).get();
        const batch = db.batch();
        existing.docs.forEach((d) => { if (d.id !== req.params.addrId) batch.update(d.ref, { is_default: false }); });
        await batch.commit();
      }
      await addrRef.update(updates);
      res.json({ message: "Address updated" });
    } catch (err) {
      res.status(400).json({ error: String(err) });
    }
  }
);

/** DELETE /api/customers/me/addresses/:addrId — remove address */
customersRouter.delete(
  "/me/addresses/:addrId",
  requireAuth,
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const customerDoc = await findMyCustomer(req.user!.uid, req.user!.email);
      if (!customerDoc) { res.status(404).json({ error: "No customer profile found" }); return; }
      const customerId = customerDoc.id;
      const addrRef = db.collection("customers").doc(customerId)
        .collection("addresses").doc(req.params.addrId);
      const addrDoc = await addrRef.get();
      if (!addrDoc.exists) { res.status(404).json({ error: "Address not found" }); return; }
      await addrRef.delete();
      res.json({ message: "Address removed" });
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);

/** GET /api/customers/:id — full profile (admin) */
customersRouter.get(
  "/:id",
  requireAuth,
  requirePermission("VIEW_CUSTOMERS"),
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const doc = await db.collection("customers").doc(req.params.id).get();
      if (!doc.exists) { res.status(404).json({ error: "Customer not found" }); return; }

      const [addresses, billing, themes, ratings, subs] = await Promise.all([
        db.collection("customers").doc(req.params.id).collection("addresses").get(),
        db.collection("customers").doc(req.params.id).collection("billing_addresses").get(),
        db.collection("customer_themes").where("customer_id", "==", req.params.id).get(),
        db.collection("customer_content_ratings").where("customer_id", "==", req.params.id).get(),
        db.collection("subscriptions").where("customer_id", "==", req.params.id)
          .orderBy("start_date", "desc").limit(20).get(),
      ]);

      res.json({
        id: doc.id,
        ...doc.data(),
        addresses:                  addresses.docs.map((d) => ({ id: d.id, ...d.data() })),
        billing_addresses:          billing.docs.map((d) => ({ id: d.id, ...d.data() })),
        theme_preferences:          themes.docs.map((d) => d.data()),
        content_rating_preferences: ratings.docs.map((d) => d.data()),
        subscriptions:              subs.docs.map((d) => ({ id: d.id, ...d.data() })),
      });
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);

/** POST /api/customers — create */
customersRouter.post(
  "/",
  requireAuth,
  requirePermission("EDIT_CUSTOMERS"),
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const { first_name, last_name, email, birth_date, accnt_theme_id } = req.body;
      if (!first_name || !last_name || !email || !birth_date || !accnt_theme_id) {
        res.status(400).json({ error: "Missing required fields: first_name, last_name, email, birth_date, accnt_theme_id" });
        return;
      }

      // Build search_index for prefix search
      const search_index = [
        ...first_name.toLowerCase().split(""),
        ...last_name.toLowerCase().split(""),
        email.toLowerCase(),
        `${first_name} ${last_name}`.toLowerCase(),
      ].filter((v, i, a) => a.indexOf(v) === i);

      const ref = await db.collection("customers").add({
        auth_uid:        req.user!.uid,
        first_name,
        last_name,
        email,
        phone_number:    req.body.phone_number    ?? null,
        birth_date,
        age_restricted:  req.body.age_restricted  ?? false,
        accnt_theme_id,
        clothing_size:   req.body.clothing_size   ?? null,
        age_rating_pref: req.body.age_rating_pref ?? "ALL",
        search_index,
        created_at: FieldValue.serverTimestamp(),
      });

      res.status(201).json({ message: "Customer created", customer_id: ref.id });
    } catch (err) {
      res.status(400).json({ error: String(err) });
    }
  }
);

/** PUT /api/customers/:id — update */
customersRouter.put(
  "/:id",
  requireAuth,
  requirePermission("EDIT_CUSTOMERS"),
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const doc = await db.collection("customers").doc(req.params.id).get();
      if (!doc.exists) { res.status(404).json({ error: "Customer not found" }); return; }

      const allowed = ["first_name", "last_name", "email", "phone_number",
                       "clothing_size", "age_rating_pref", "accnt_theme_id", "age_restricted"];
      const updates: Record<string, unknown> = {};
      for (const key of allowed) {
        if (req.body[key] !== undefined) updates[key] = req.body[key];
      }

      if (Object.keys(updates).length === 0) {
        res.status(400).json({ error: "No valid fields to update" }); return;
      }

      updates.updated_at = FieldValue.serverTimestamp();
      await db.collection("customers").doc(req.params.id).update(updates);
      await writeAudit(req.user!.uid, "customer", req.params.id, "UPDATE", "", JSON.stringify(updates));

      res.json({ message: "Customer updated" });
    } catch (err) {
      res.status(400).json({ error: String(err) });
    }
  }
);

/** DELETE /api/customers/:id */
customersRouter.delete(
  "/:id",
  requireAuth,
  requirePermission("DELETE_CUSTOMERS"),
  async (req: AuthRequest, res: Response): Promise<void> => {
    try {
      const hasSubs = await db.collection("subscriptions")
        .where("customer_id", "==", req.params.id).limit(1).get();
      if (!hasSubs.empty) {
        res.status(409).json({ error: "Cannot delete customer with active subscriptions" }); return;
      }

      await db.collection("customers").doc(req.params.id).delete();
      res.json({ message: "Customer deleted" });
    } catch (err) {
      res.status(500).json({ error: String(err) });
    }
  }
);
